src/sagemaker_xgboost_container/distributed_gpu/distributed_gpu_training.py

```diff
@@ -78,16 +111,63 @@ def run_training_with_dask(
                 watchlist.append((dvalid, "validation"))
 
             logging.info("Data load complete. Starting training...")
 
+            # Redundant Code -------------------------------------------------------------------- >
+            """
+            The blob below is the redundant code we can see in the original training flow. Some
+            points that might be worth addressing in the copied blob below are as follows:
+            * Are hardcoded static string references everywhere really the best way to do this?
+            * Can we consolidate this data cleanup and popping business into an appropriate class?
+            * Similar to the above point, is an anemic data model/similar an angle worth considering?
+            * Do we have overhead concerns with the debugging if we need to convert to DMatrix?
+            * Does allowing for cross validation outweigh overhead concerns between CPU & GPU?
+            """
+            num_round = hyperparameters.pop("num_round")
+            save_model_on_termination = hyperparameters.pop("save_model_on_termination", "false")
+            tuning_objective_metric_param = hyperparameters.pop("_tuning_objective_metric", None)
+            eval_metric = hyperparameters.pop("eval_metric", None)
+            cleaned_eval_metric, configured_feval, tuning_objective_metric = train_utils.get_eval_metrics_and_feval(
+                tuning_objective_metric_param, eval_metric
+            )
+            if cleaned_eval_metric:
+                hyperparameters["eval_metric"] = cleaned_eval_metric
+
+            early_stopping_data_name = "validation" if dvalid else None
+            early_stopping_rounds = hyperparameters.pop("early_stopping_rounds", None)
+
+            early_stopping_metric = None
+            if early_stopping_rounds:
+                if tuning_objective_metric:
+                    early_stopping_metric = tuning_objective_metric[-1]
+                elif eval_metric:
+                    early_stopping_metric = eval_metric[-1]
+
+            xgb_model, iteration, callbacks = get_callbacks(
+                model_dir=model_dir,
+                checkpoint_dir=checkpoint_dir,
+                early_stopping_data_name=early_stopping_data_name,
+                early_stopping_metric=early_stopping_metric,
+                early_stopping_rounds=early_stopping_rounds,
+                save_model_on_termination=save_model_on_termination,
+                is_master=is_scheduler_host,
+            )
+
             try:
                 output = xgb.dask.train(
-                    client, hyperparameters, dtrain, num_boost_round=hyperparameters["num_round"], evals=watchlist
+                    client=client,
+                    params=hyperparameters,
+                    dtrain=dtrain,
+                    num_boost_round=num_round,
+                    evals=watchlist,
+                    feval=configured_feval,
```

Comment:
Note that I've found out that this custom metric arg increases the training time by considerable amount for training with large data. It's used when `eval_metric` is specified by the user or tuning metric is specified for HPO. Looking at the underlying code it deals with DMatrix directly, which could be the reason for the slowdown. It doesn't look like we're doing anything wrong.